src/mysql_k8s_helpers.py

```diff
@@ -267,78 +267,83 @@ def execute_backup_commands(
     ) -> Tuple[str, str]:
         """Executes the run_backup.sh script in the container with the given args."""
-        nproc_commands = "nproc".split()
+        nproc_command = "nproc".split()
 
-        make_temp_dir_commands = "mktemp --tmpdir --directory xtra_backup_XXXX".split()
+        make_temp_dir_command = "mktemp --tmpdir --directory xtra_backup_XXXX".split()
 
         try:
-            process = self.container.exec(nproc_commands)
+            process = self.container.exec(nproc_command)
             nproc, _ = process.wait_output()
 
-            process = self.container.exec(make_temp_dir_commands)
+            process = self.container.exec(make_temp_dir_command)
             tmp_dir, _ = process.wait_output()
         except ExecError as e:
             logger.exception("Failed to execute commands prior to running backup", exc_info=e)
             raise MySQLExecuteBackupCommandsError(e.stderr)
+        except Exception as e:
+            # Catch all other exceptions to prevent the database being stuck in
+            # a bad state due to pre-backup operations
+            logger.exception("Failed to execute commands prior to running backup", exc_info=e)
+            raise MySQLExecuteBackupCommandsError(e)
 
-        # TODO: remove flags --no-version-check and --no-server-version-check
+        # TODO: remove flags --no-server-version-check
         # when MySQL and XtraBackup versions are in sync
         xtrabackup_commands = " ".join(
-            """
+            f"""
 xtrabackup --defaults-file=/etc/mysql
             --defaults-group=mysqld
             --no-version-check
-            --parallel="$NPROC"
-            --user="$MYSQL_USER"
+            --parallel={nproc.strip()}
+            --user="{self.server_config_user}"
             --password="$MYSQL_PASSWORD"
             --socket=/run/mysqld/mysqld.sock
             --lock-ddl
             --backup
             --stream=xbstream
             --xtrabackup-plugin-dir=/usr/lib64/xtrabackup/plugin
-            --target-dir="$TMP_DIRECTORY"
+            --target-dir="{tmp_dir.strip()}"
             --no-server-version-check
     | xbcloud put
             --curl-retriable-errors=7
             --insecure
             --storage=s3
             --parallel=10
             --md5
-            --s3-bucket="$S3_BUCKET"
-            --s3-access-key="$S3_ACCESS_KEY"
-            --s3-secret-key="$S3_SECRET_KEY"
-            "$S3_PATH"
-""".strip().split()
+            --s3-bucket="{s3_bucket}"
+            --s3-endpoint="{s3_endpoint}"
+            "{s3_directory}"
+""".split()
         )
-        # TODO: run xtrabackup_commands directly when pebble supports
-        # env var expansion (https://github.com/canonical/pebble/issues/187)
-        backup_commands = ["sh", "-c", f"{xtrabackup_commands}"]
+        # Use sh to be able to use the pipe in above commands
+        backup_commands = ["sh", "-c", f"{{ read ''MYSQL_PASSWORD; {xtrabackup_commands}; }}"]
```

Comment:
pass stdin directly to xtrabackup. reason: xtrabackup will read the password from stdin if the password parameter is empty.
```suggestion
        backup_commands = ["sh", "-c", f"{xtrabackup_commands}"]
```
it is ok to handle this in a separate ticket/PR

Reply:
addressed in 2f1db018fed3b71ac85361e7b9ca380e0cc7c95a